File: mystore/orders/payment.py
import dataclasses
import os

import stripe
from django.conf import settings
from django.utils.http import urlsafe_base64_encode
from rest_framework import status
from rest_framework.response import Response

# def update_carrier(self, charge, customer, carrier, tracking_number):
#     try:
#         response = stripe.PaymentIntent.modify(
#             charge,
#             customer=customer,
#             shipping={
#                 'carrier': carrier,
#                 'tracking_number': tracking_number,
#             }
#         )
#     except stripe.StripeError as e:
#         self.errors['update_error'] = e.args
#     except Exception as e:
#         self.errors['update_error'] = e.args
#     else:
#         return True
#     finally:
#         if self.errors:
#             return False


@dataclasses.dataclass
class PaymentDetails:
    payment_intent_id: str = None
    client_secret: str = None


class StripeInterfaceMixin:
    def __init__(self):
        self.completed = False
        self.response_data = {}
        self.errors = {}

        if settings.DEBUG:
            stripe.api_key = os.getenv('STRIPE_TEST_SECRET_KEY')
        else:
            stripe.api_key = os.getenv('STRIPE_PRODUCTION_API_KEY')

    @staticmethod
    def adapt_number(value):
        return int(float(value)) * 100


class PaymentInterface(StripeInterfaceMixin):
    def __init__(self):
        self.payment_details = PaymentDetails()
        super().__init__()

    def get_fail_response(self, message=None, **kwargs):
        data = {'message': 'Payment failed'} | kwargs | self.errors
        return Response(data, status=status.HTTP_402_PAYMENT_REQUIRED)

    def get_success_response(self, message=None, **kwargs):
        data = {
            'intent': self.payment_details.payment_intent_id,
            'client': self.payment_details.client_secret
        }
        if message is not None:
            data['message'] = message
        self.response_data = data | kwargs
        return Response(self.response_data, status=status.HTTP_201_CREATED)

    def create_new_source(self, customer, source):
        try:
            response = stripe.Customer.create_source(
                customer,
                source=source
            )
        except stripe.StripeError as e:
            self.errors['payment_error'] = e.args
        except Exception as e:
            self.errors['payment_error'] = e.args
        finally:
            if self.errors:
                return False
        return response

    def payment_intent(self, request, amount, billing_adress, debug=False, products=[], stripe_params={}):
        session_id = stripe_params.get('session_id')

        params = {
            'amount': self.adapt_number(amount),
            'setup_future_usage': 'off_session',
            'payment_method_types': ['card'],
            'idempotency_key': session_id,
            'automatic_payment_methods': {
                'enabled': False
            },
            'currency': 'eur',
            'customer': request.user.userprofile.stripe_id,
            'description': 'Customer order for products',
            'metadata': {},
            'shipping': {
                'address': {
                    'line1': getattr(billing_adress, 'address_line'),
                    'line2': None,
                    'postal_code': getattr(billing_adress, 'zip_code'),
                    'city': getattr(billing_adress, 'city'),
                    'state': None,
                    'country': getattr(billing_adress, 'country'),
                },
                'name': billing_adress.get_full_name,
                'phone': getattr(billing_adress, 'telephone'),
                'carrier': None
            },
            'receipt_email': request.user.email,
        }

        try:
            response = stripe.PaymentIntent.create(**params)
        except stripe.StripeError as e:
            self.errors['payment_error'] = e.args
        except Exception as e:
            self.errors['payment_error'] = e.args
        finally:
            if self.errors:
                return False

        intent_id = response['id']
        client_secret = response['client_secret'].encode('utf-8')
        token = urlsafe_base64_encode(bytes(client_secret))

        self.payment_details.client_secret = token
        self.payment_details.payment_intent_id = intent_id
        self.completed = True
        return True

    def capture_intent(self, request, intent, payment_method):
        try:
            response = stripe.PaymentIntent.confirm(
                intent,
                payment_method=payment_method
            )
        except stripe.StripeError as e:
            self.errors['payment_error'] = e.args
        except Exception as e:
            self.errors['payment_error'] = e.args
        finally:
            if self.errors:
                return False
                        
        self.completed = True
        return response

# def refund(self, request, charge, reason=None):
#        try:
#             response = stripe.Refund.create(
#                 charge=charge,
#                 reason=None,
#                 metadata={}
#             )
#         except stripe.StripeError as e:
#             self.errors['update_error'] = e.args
#         except Exception as e:
#             self.errors['update_error'] = e.args
#         else:
#             return True
#         finally:
#             if self.errors:
#                 return False

File: services/gopurchase/internal/handlers/payment.go
package handlers

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"

	"github.com/Zadigo/gopurchase/internal/utils"
	"github.com/stripe/stripe-go/v85"
)

// SetupStripeClient initializes the Stripe client with the API key
// from the environment variable.
func (p *PaymentApi) SetupStripeClient() error {
	key := os.Getenv("STRIPE_API_KEY")
	if key == "" {
		return fmt.Errorf("STRIPE_API_KEY environment variable is not set")
	}
	p.PaymentClient = stripe.NewClient(key)
	return nil
}

func (p *PaymentApi) CreateIntent(w http.ResponseWriter, r *http.Request) {
	// Try to get an active customer from the shop database. We should receeive
	// a customer ID that we can use to create the payment intent since Django
	// automatically creates a Stripe customer for each user that registers
	// on the platform. If a customer does not exist, we will be able to update
	// it later in the UpdateIntent method once the customer creates his account.
	// var customerID string

	// requestData, _ := json.Marshal(map[string]any{"email": "something@gmail.com"})
	// reader := bytes.NewReader(requestData)

	// responseData := map[string]any{"customerId": ""}
	// err := utilities.SendRequest("https://example.com/", reader, responseData)

	// if err != nil {
	// 	// Do something
	// } else {
	// 	if id, ok := responseData["customerId"].(string); ok {
	// 		customerID = id
	// 	}
	// }

	errorHandler := NewHttpErrorHandler(w)

	data := CreatePaymentIntentRequest{}
	err := json.NewDecoder(r.Body).Decode(&data)
	if err != nil {
		message := utils.DefaultErrorResponse{Detail: "Invalid request payload", Message: err.Error()}
		utils.JsonResponse(w, message, http.StatusBadRequest)
		return
	}

	// TODO: Session ID should be JWT token in order to be able to verify
	// its authenticity and prevent malicious users from creating payment intents
	// with random session IDs. The JWT token should contain the session ID and any other
	// relevant information that we might need to create the payment intent. We can then verify the
	// JWT token using a secret key and extract the session ID from it to create the payment intent.
	if data.SessionId == "" {
		errorHandler.InvalidSessionIdError(err)
		return
	}

	options := &stripe.PaymentIntentCreateParams{
		// Customer:    stripe.String(customerID),
		// ReturnURL:   stripe.String("https://example.com/return_url"),
		// PaymentMethod: stripe.String("pm_card_visa"),
		// Set a default amount of 0.50 EUR for creation. This will be overridden during
		// the cusomter's checkout process when the total amount is calculated.
		Amount:           stripe.Int64(0.50 * 100), // Convert to cents
		Currency:         stripe.String(string(stripe.CurrencyEUR)),
		Description:      stripe.String("Test Payment Intent"),
		Metadata:         map[string]string{"sessionId": data.SessionId},
		SetupFutureUsage: stripe.String("off_session"),
		AmountDetails: &stripe.PaymentIntentCreateAmountDetailsParams{
			LineItems: data.Items.CreateLineItems(),
		},
		PaymentMethodTypes: []*string{
			stripe.String("card"),
		},
	}

	if data.Total > 0 {
		// Only set the amount if it's greater than 0 in order to avoid
		// raising an error from Stripe since the amount is required when
		// creating a payment intent.
		options.Amount = stripe.Int64(int64(data.Total * 100)) // Convert to cents
	}

	intent, err := p.PaymentClient.V1PaymentIntents.Create(p.Ctx, options)
	if err != nil {
		log.Printf("❌ Failed to create payment intent: %v", err)
		errorHandler.PaymentIntentCreationError(err)
		return
	}

	// Store the payment intent in Redis
	redisHandler := NewPaymentRedis(p.App.GetRedisClient())
	err = redisHandler.SetPaymentIntent(intent)
	if err != nil {
		log.Printf("❌ Failed to store payment intent in Redis: %v", err)
	}

	responseData := map[string]string{
		"paymentIntentId": intent.ID,
		"message":         "Payment intent created successfully",
	}

	utils.JsonResponse(w, responseData, http.StatusOK)
}

func (p *PaymentApi) UpdateIntent(w http.ResponseWriter, r *http.Request) {
	errorHandler := NewHttpErrorHandler(w)

	data := UpdatePaymentIntentRequest{}
	err := json.NewDecoder(r.Body).Decode(&data)
	if err != nil {
		errorHandler.InvalidBodyError(err)
		return
	}

	if data.PaymentIntentData.PaymentIntentID == "" {
		errorHandler.PaymentIntentMissingError(err)
		return
	}

	params := &stripe.PaymentIntentUpdateParams{
		ReceiptEmail: stripe.String(data.Email),
		AmountDetails: &stripe.PaymentIntentUpdateAmountDetailsParams{
			LineItems: data.Items.UpdateLineItems(),
		},
		Shipping: &stripe.ShippingDetailsParams{
			Name:  stripe.String(fmt.Sprintf("%s %s", data.Firstname, data.Lastname)),
			Phone: stripe.String(data.Telephone),
			Address: &stripe.AddressParams{
				City:       stripe.String(data.City),
				Country:    stripe.String(data.Country),
				Line1:      stripe.String(data.AddressLine),
				Line2:      stripe.String(""),
				PostalCode: stripe.String(data.PostalCode),
				State:      stripe.String(data.State),
			},
		},
	}

	if data.CustomerID != "" {
		params.Customer = stripe.String(data.CustomerID)
	}

	intent, err := p.PaymentClient.V1PaymentIntents.Update(p.Ctx, data.PaymentIntentID, params)

	if err != nil {
		errorHandler.PaymentIntentUpdateError(err)
		return
	}

	redisHandler := NewPaymentRedis(p.App.GetRedisClient())
	err = redisHandler.UpdatePaymentIntent(intent)
	if err != nil {
		log.Printf("❌ Failed to update payment intent in Redis: %v", err)
	}

	responseData := map[string]string{
		"paymentIntentId": intent.ID,
		"message":         "Payment intent updated successfully",
	}

	utils.JsonResponse(w, responseData, http.StatusOK)
}

func (p *PaymentApi) CaptureIntent(w http.ResponseWriter, r *http.Request) {
	errorHandler := NewHttpErrorHandler(w)

	data := CapturePaymentIntentRequest{}
	err := json.NewDecoder(r.Body).Decode(&data)
	if err != nil {
		errorHandler.InvalidBodyError(err)
		return
	}

	intent, err := p.PaymentClient.V1PaymentIntents.Confirm(p.Ctx, data.PaymentIntentID, &stripe.PaymentIntentConfirmParams{
		ReturnURL:     stripe.String("https://example.com/return_url"),
		// PaymentMethod: stripe.String("pm_card_mastercard"),
		PaymentMethod: stripe.String(data.Card),
	})

	if err != nil {
		errorHandler.PaymentIntentCaptureError(err)
		return
	}

	redisHandler := NewPaymentRedis(p.App.GetRedisClient())
	err = redisHandler.UpdatePaymentIntent(intent)
	if err != nil {
		log.Printf("❌ Failed to update payment intent in Redis: %v", err)
	}

	responseData := map[string]string{
		"paymentIntentId": intent.ID,
		"message":         "Payment intent captured successfully",
	}

	utils.JsonResponse(w, responseData, http.StatusOK)
}
